### PR TITLE
Add function to read entity sets JSON file in IO

### PR DESCRIPTION
**Pull request by tianchiTJ**

### CODE REVIEW

File: include/io.h [hunk 1]
```diff
@@ -11,6 +11,8 @@
 //! Alias for JSON
 #include "json.hpp"
 using Json = nlohmann::json;
+//! MPM Index
```
Comment (kks32) [inline include/io.h]: Please define it inside the namespace `mpm`.

File: include/io.h [hunk 2]
```diff
@@ -51,6 +53,13 @@ class IO {
   //! Return json object
   Json json_object(const std::string& name) const { return json_[name]; }
 
+  //! Return the entity sets from the input set JSON file
+  //! \param[in] filename File name
+  //! \param[in] sets_type type of sets
+  //! \return map of entity sets
```
Comment (kks32) [inline include/io.h]:
```suggestion
  //! \retval entity_sets map of entity sets
```

File: src/io.cc [hunk 3]
```diff
@@ -161,3 +161,43 @@ std::string mpm::IO::output_folder() const {
...
+  sfs.exceptions(std::ifstream::failbit | std::ifstream::badbit);
+  // Map of the entity sets
+  std::map<Index, std::vector<Index>> entity_sets;
+
+  try {
+    sfs.open(filename);
+    // Entity sets JSON object
+    Json json_sets_ = Json::parse(sfs);
+    // type of sets (ex: node sets or particle sets)
+    Json esets = json_sets_["sets_type"];
```
Comment (kks32) [inline src/io.cc]: Could we rename `esets` to `sets`

Comment (kks32) [inline src/io.cc]:
```suggestion
    Json sets = json_sets_[sets_type];
```

Comment (kks32) [reply src/io.cc]: `sets_type` variable can't be inside a string ""

File: src/io.cc [hunk 4]
```diff
@@ -161,3 +161,43 @@ std::string mpm::IO::output_folder() const {
   }
   return path;
 }
+
+//! Return map of entity sets from the JSON file
+std::map<Index, std::vector<Index>> mpm::IO::entity_sets(
+    const std::string& filename, const std::string& sets_type) {
+
+  // Input file stream for sets JSON file
+  std::ifstream sfs;
```
Comment (kks32) [inline src/io.cc]: Could you rename `sfs` to `sets_file`

File: src/io.cc [hunk 5]
```diff
@@ -161,3 +161,43 @@ std::string mpm::IO::output_folder() const {
...
+  sfs.exceptions(std::ifstream::failbit | std::ifstream::badbit);
+  // Map of the entity sets
+  std::map<Index, std::vector<Index>> entity_sets;
+
+  try {
+    sfs.open(filename);
+    // Entity sets JSON object
+    Json json_sets_ = Json::parse(sfs);
+    // type of sets (ex: node sets or particle sets)
```
Comment (kks32) [inline src/io.cc]:
```suggestion
    // type of sets (e.g: node_sets or particle_sets)
```

File: src/io.cc [hunk 6]
```diff
@@ -161,3 +161,43 @@ std::string mpm::IO::output_folder() const {
...
+    Json esets = json_sets_["sets_type"];
+
+    if (esets.size() > 0) {
+      for (Json::iterator itr = esets.begin(); itr != esets.end(); ++itr) {
+        // Get the entity set ids
+        Index id = (*itr)["id"].template get<Index>();
+        // Get the entity ids
+        std::vector<Index> eids;
```
Comment (kks32) [inline src/io.cc]:
```suggestion
        std::vector<Index> entity_ids;
```

Comment (kks32) [reply src/io.cc]: Rename to `entity_ids` or `entities`

File: src/io.cc [hunk 7]
```diff
@@ -161,3 +161,43 @@ std::string mpm::IO::output_folder() const {
...
+        // Get the size of the set vector
+        eids.resize((*itr).at("set").size());
+
+        for (unsigned i = 0; i < (*itr).at("set").size(); ++i)
+          eids.emplace_back((*itr).at("set").at(i).template get<Index>());
+
+        entity_sets[id] = eids;
```
Comment (kks32) [inline src/io.cc]: I thin map has an insert function, which returns a boolean, could we try to use that?

**Review by kks32 — CHANGES_REQUESTED** [review]

**Review by kks32 — COMMENTED** [review]
Excellent work! Just a few minor tweaks

File: src/io.cc (hunk 3 above)
Comment (tianchiTJ) [reply src/io.cc]: Sure

File: src/io.cc (hunk 7 above)
Comment (kks32) [reply src/io.cc]: http://www.cplusplus.com/reference/map/map/insert/

File: src/io.cc (hunk 4 above)
Comment (tianchiTJ) [reply src/io.cc]: Ok

File: src/io.cc (hunk 7 above)
Comment (tianchiTJ) [reply src/io.cc]: Ok, I will

File: src/io.cc [hunk 8]
```diff
@@ -161,3 +161,46 @@ std::string mpm::IO::output_folder() const {
...
+        for (unsigned i = 0; i < (*itr).at("set").size(); ++i)
+          entity_ids.emplace_back(
+              (*itr).at("set").at(i).template get<mpm::Index>());
+        // Add the entity set to the list
+        entity_sets.insert(
+            std::pair<mpm::Index, std::vector<mpm::Index>>(id, entity_ids));
+      }
+    }
+    sets_file.close();
+  } catch (const std::exception& exception) {
```
Comment (kks32) [inline src/io.cc:197]:
```suggestion
const std::out_of_range& range_error) {
        console_->warn("{} {} reading {}: {}", __FILE__, __LINE__, sets_type, filename,
                    range_error.what())
 }
catch (const std::exception& exception) {
  console_->error("{} #{}: {}\n", __FILE__, __LINE__, exception.what());
}
```

**Review by kks32 — CHANGES_REQUESTED** [review]

File: src/io.cc [hunk 9]
```diff
@@ -161,3 +161,46 @@ std::string mpm::IO::output_folder() const {
...
+    Json json_sets_ = Json::parse(sets_file);
+    // type of sets (e.g: node_sets or particle_sets)
+    Json sets = json_sets_[sets_type];
+
+    if (sets.size() > 0) {
+      for (Json::iterator itr = sets.begin(); itr != sets.end(); ++itr) {
+        // Get the entity set ids
+        mpm::Index id = (*itr)["id"].template get<mpm::Index>();
+        // Get the entity ids
+        std::vector<mpm::Index> entity_ids;
```
Comment (kks32) [inline src/io.cc]:
```suggestion
 std::vector<mpm::Index> entity_ids = (*itr).at("set");
```

Comment (kks32) [reply src/io.cc]: Lines  187 - 193 is not needed.

**Review by kks32 — APPROVED** [review]

**Pull request merged**